cmake/toolchain/x86_64.cmake

```diff
@@ -0,0 +1,47 @@
+#
+# Copyright (C) 2020 Assured Information Security, Inc.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+set(CMAKE_SYSTEM_NAME Generic)
+set(CMAKE_CXX_COMPILER_WORKS 1)
+
+string(CONCAT HYPERVISOR_CXX_FLAGS
+    "--target=x86_64-elf "
+    "-ffreestanding "
+    "-mno-red-zone "
+    "-D__bareflank__ "
+)
+
+string(CONCAT HYPERVISOR_LINK_FLAGS
+    "-static "
+    "-nostdlib "
+    "-z max-page-size=0x1000 "
+    "-z noexecstack "
+)
+
+set(CMAKE_ASM_COMPILE_OBJECT
+    "<CMAKE_CXX_COMPILER> ${HYPERVISOR_CXX_FLAGS} <DEFINES> <INCLUDES> <FLAGS> -o <OBJECT> -c <SOURCE>")
+
+set(CMAKE_CXX_COMPILE_OBJECT
+    "<CMAKE_CXX_COMPILER> ${HYPERVISOR_CXX_FLAGS} <DEFINES> <INCLUDES> <FLAGS> -o <OBJECT> -c <SOURCE>")
+
+set(CMAKE_CXX_LINK_EXECUTABLE
```

Comment:
Why couldn't you have just added this CMAKE_CXX_LINK_EXECUTABLE command to the existing generic toolchain? I feel like you could have left all of these flags at the cmake targets they belong to, updated the --target options where they used to exist, and then not used external project add?

The only thing this toolchain really provides is the ability to link vmm binaries with lld, while not using lld for userspace binaries. Is there some reason that you explicitly don't want to use lld to link the userspace binaries?

Reply:
If you set the link command it would change linking for everything, including the native compiled parts. The issue with the triple that we need is it will not use the correct linker, and there is no way to change that without either using EPA or making a mod to clang

Reply:
Right, this sets the link command template for everything in a given cmake project. So what I'm asking is if the following works:
* Remove the HYPERVISOR_CXX_FLAGS and HYPERVISOR_LINK_FLAGS flags from this toolchain file (moving them back to the cmake targets)
* Keep HYPERVISOR_CXX_LINKER in this toolchain file's link command template (which would force cmake to link with lld directly regardless of the clang --target option specified)
* Update the clang --target options to be "more correct" at each individual cmake target (for example, specify --target=x86_64-elf in the kernel and runtime components)